python/GafferImageTest/ImageReaderTest.py

```diff
@@ -463,5 +463,48 @@ def testDisabling( self ) :
 
 		self.assertImagesEqual( reader["out"], constant["out"] )
 
+	def testLayerNames( self ):
+
+		reader = GafferImage.ImageReader()
+		def validateChannels( channels ):
+			self.assertEqual( [ i for i in reader["out"].channelNames() ], [ i[0] for i in channels ] )
+			for name, value in channels:
+				self.assertAlmostEqual( reader["out"].channelData( name, imath.V2i( 0 ) )[0], value, places = 3 )
+			
+
+		reader["fileName"].setValue( "${GAFFER_ROOT}/python/GafferImageTest/images/imitateProductionLayers1.exr" )
+
+		reader["channelNaming"].setValue( GafferImage.OpenImageIOReader.ChannelNaming.Default )
+		validateChannels( [ ( "R", 0.1 ), ( "G", 0.2 ), ( "B", 0.3 ), ( "character.R", 0.4 ), ( "character.G", 0.5 ), ( "character.B", 0.6 ) ] )
+		reader["channelNaming"].setValue( GafferImage.OpenImageIOReader.ChannelNaming.Legacy )
+		validateChannels( [ ( "rgba.main.R", 0.1 ), ( "rgba.main.G", 0.2 ), ( "rgba.main.B", 0.3 ), ( "character.main.red", 0.4 ), ( "character.main.green", 0.5 ), ( "character.main.blue", 0.6 ) ] )
+		reader["channelNaming"].setValue( GafferImage.OpenImageIOReader.ChannelNaming.Specification )
+		validateChannels( [ ( "R", 0.1 ), ( "G", 0.2 ), ( "B", 0.3 ), ( "red", 0.4 ), ( "green", 0.5 ), ( "blue", 0.6 ) ] )
+
+		reader["fileName"].setValue( "${GAFFER_ROOT}/python/GafferImageTest/images/imitateProductionLayers2.exr" )
+		reader["channelNaming"].setValue( GafferImage.OpenImageIOReader.ChannelNaming.Default )
+		validateChannels( [ ( "R", 0.1 ), ( "G", 0.2 ), ( "B", 0.3 ), ( "Z", 0.4 ) ] )
+		reader["channelNaming"].setValue( GafferImage.OpenImageIOReader.ChannelNaming.Legacy )
+		validateChannels( [ ( "rgba_main.R", 0.1 ), ( "rgba_main.G", 0.2 ), ( "rgba_main.B", 0.3 ), ( "depth_main.depth.Z", 0.4 ) ] )
+		reader["channelNaming"].setValue( GafferImage.OpenImageIOReader.ChannelNaming.Specification )
+		validateChannels( [ ( "R", 0.1 ), ( "G", 0.2 ), ( "B", 0.3 ), ( "depth.Z", 0.4 ) ] )
+
+		reader["fileName"].setValue( "${GAFFER_ROOT}/python/GafferImageTest/images/weirdPartNames.exr" )
+		reader["channelNaming"].setValue( GafferImage.OpenImageIOReader.ChannelNaming.Default )
+		validateChannels( [
+			( "part0.R", 0.1 ), ( "part1.G", 0.2 ), ( "part2.B", 0.3 ), ( "layer.R", 0.4 ),
+			( "part3.A", 0.4 ), ( "layer.G", 0.5 ), ( "layer.B", 0.6 ), ( "layer.A", 1.0 )
+		] )
```

Comment:
I have been unable to devise a heuristic that would differentiate between "part: diffuse, channel: R", which is how Nuke writes "diffuse.R", and "part: part0, channel: R" in this situation.  We usually avoid this in practice because we notice that the main part is named "main" or "rgb" or "RGBA" or "depth", but according to the spec, the main part can easily be named "part0" or anything else, and I don't think there's any way we can figure this out.  The heuristic mode will definitely not be ironclad, but it has dealt with all the production cases I've thrown at it so far.

Reply:
Ah, so this `part0, part1...` convention isn't one we've encountered in the wild, but is just something we've added for testing? Seems a pity to add something that thwarts us artificially, though I can understand wanting coverage for the extra-weird cases too. What would you say to a heuristic that says "`partN` is just a part name, and definitely not a layer name"? Too much risk that someone would call a layer `part1` in Nuke?